<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork900-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


<h1>Extracting and Visualizing Stock Data</h1>
<h2>Description</h2>


Extracting essential data from a dataset and displaying it is a necessary part of data science; therefore individuals can make correct decisions based on the data. In this assignment, you will extract some stock data, you will then display this data in a graph.


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>Define a Function that Makes a Graph</li>
        <li>Question 1: Use yfinance to Extract Stock Data</li>
        <li>Question 2: Use Webscraping to Extract Tesla Revenue Data</li>
        <li>Question 3: Use yfinance to Extract Stock Data</li>
        <li>Question 4: Use Webscraping to Extract GME Revenue Data</li>
        <li>Question 5: Plot Tesla Stock Graph</li>
        <li>Question 6: Plot GameStop Stock Graph</li>
    </ul>
<p>
    Estimated Time Needed: <strong>30 min</strong></p>
</div>

<hr>


***Note***:- If you are working Locally using anaconda, please uncomment the following code and execute it.


In [100]:
##!pip install yfinance==0.2.38
##!pip install pandas==2.2.2
##!pip install nbformat

In [101]:
!pip install yfinance==0.1.67
!mamba install bs4==4.10.0 -y
!pip install nbformat==4.2.0


                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (1.4.2) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████


Looking for: ['bs4==4.10.0']

[+] 0.0s
pkgs/main/linux-64 ━╸━━━━━━━━━━━━━━━╸━━━━━━━   0.0 B /  ??.?MB @  ??.?MB/s

In [102]:
import yfinance as yf
import pandas as pd
import requests
from bs4 import BeautifulSoup
import plotly.graph_objects as go
from plotly.subplots import make_subplots

In Python, you can ignore warnings using the warnings module. You can use the filterwarnings function to filter or ignore specific warning messages or categories.


In [103]:
import warnings
# Ignore all warnings
warnings.filterwarnings("ignore", category=FutureWarning)

## Define Graphing Function


In this section, we define the function `make_graph`. You don't have to know how the function works, you should only care about the inputs. It takes a dataframe with stock data (dataframe must contain Date and Close columns), a dataframe with revenue data (dataframe must contain Date and Revenue columns), and the name of the stock.


In [104]:
def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing = .3)
    stock_data_specific = stock_data[stock_data.Date <= '2021--06-14']
    revenue_data_specific = revenue_data[revenue_data.Date <= '2021-04-30']
    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date, infer_datetime_format=True), y=stock_data_specific.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date, infer_datetime_format=True), y=revenue_data_specific.Revenue.astype("float"), name="Revenue"), row=2, col=1)
    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)
    fig.update_layout(showlegend=False,
    height=900,
    title=stock,
    xaxis_rangeslider_visible=True)
    fig.show()

## Question 1: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is Tesla and its ticker symbol is `TSLA`.


In [105]:
import yfinance as yf

# Creating a ticker object for Tesla
tesla_ticker = yf.Ticker("TSLA")

# Display the ticker object to confirm it has been created
print(tesla_ticker)

yfinance.Ticker object <TSLA>


Using the ticker object and the function `history` extract stock information and save it in a dataframe named `tesla_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [106]:
ticker = yf.Ticker("TSLA")

tesla_data = ticker.history(period="max")

print("Tesla stock data extracted and saved in tesla_data dataframe")

Tesla stock data extracted and saved in tesla_data dataframe


**Reset the index** using the `reset_index(inplace=True)` function on the tesla_data DataFrame and display the first five rows of the `tesla_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 1 to the results below.


In [107]:
import requests
from bs4 import BeautifulSoup
import pandas as pd
from yfinance import Ticker

url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/amazon_data_webpage.html"
response = requests.get(url)
html_data = response.text
soup = BeautifulSoup(html_data, 'html.parser')
amazon_data = pd.read_html(url)[0]

ticker_symbol = "TSLA"
ticker = Ticker(ticker_symbol)
tesla_data = ticker.history(period="max")
tesla_data.reset_index(inplace=True)

print(tesla_data.head())

        Date      Open      High       Low     Close     Volume  Dividends  \
0 2010-06-29  1.266667  1.666667  1.169333  1.592667  281494500          0   
1 2010-06-30  1.719333  2.028000  1.553333  1.588667  257806500          0   
2 2010-07-01  1.666667  1.728000  1.351333  1.464000  123282000          0   
3 2010-07-02  1.533333  1.540000  1.247333  1.280000   77097000          0   
4 2010-07-06  1.333333  1.333333  1.055333  1.074000  103003500          0   

   Stock Splits  
0           0.0  
1           0.0  
2           0.0  
3           0.0  
4           0.0  


## Question 2: Use Webscraping to Extract Tesla Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm Save the text of the response as a variable named `html_data`.


In [108]:
import requests

url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm"
response = requests.get(url)
html_data = response.text


Parse the html data using `beautiful_soup`.


In [109]:
from bs4 import BeautifulSoup

soup = BeautifulSoup(html_data, 'html.parser')

Using `BeautifulSoup` or the `read_html` function extract the table with `Tesla Revenue` and store it into a dataframe named `tesla_revenue`. The dataframe should have columns `Date` and `Revenue`.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1

We are focusing on quarterly revenue in the lab.
```

</details>


In [110]:
import pandas as pd

# Using read_html function
tesla_revenue = pd.read_html(url, header=0, index_col=0)[1]

# Alternatively, using BeautifulSoup
# soup = BeautifulSoup(html_data, 'html.parser')
# table = soup.find_all("tbody")[1]
# tesla_revenue = pd.DataFrame([[td.text for td in row.find_all("td")] for row in table.find_all("tr")], columns=["Date", "Revenue"])

print(tesla_revenue)

                                          Tesla Quarterly Revenue(Millions of US $).1
Tesla Quarterly Revenue(Millions of US $)                                            
2022-09-30                                                                    $21,454
2022-06-30                                                                    $16,934
2022-03-31                                                                    $18,756
2021-12-31                                                                    $17,719
2021-09-30                                                                    $13,757
2021-06-30                                                                    $11,958
2021-03-31                                                                    $10,389
2020-12-31                                                                    $10,744
2020-09-30                                                                     $8,771
2020-06-30                                            

Execute the following line to remove the comma and dollar sign from the `Revenue` column. 


In [122]:
import pandas as pd

tesla_revenue['Revenue'] = tesla_revenue['Revenue'].str.replace(',', '').str.replace('$', '')


KeyError: 'Revenue'

Execute the following lines to remove an null or empty strings in the Revenue column.


In [115]:
import pandas as pd

# Sample data for demonstration (replace this with your actual data loading step)
# tesla_revenue = pd.read_csv('tesla_revenue.csv')

# Print the original DataFrame (optional)
print("Original DataFrame:")
print(tesla_revenue.head())

# Remove null values in the 'Revenue' column
tesla_revenue.dropna(subset=['Revenue'], inplace=True)

# Remove empty strings in the 'Revenue' column
tesla_revenue = tesla_revenue[tesla_revenue['Revenue'] != ""]

# Print the cleaned DataFrame (optional)
print("Cleaned DataFrame:")
print(tesla_revenue.head())

Original DataFrame:
                                          Tesla Quarterly Revenue(Millions of US $).1
Tesla Quarterly Revenue(Millions of US $)                                            
2022-09-30                                                                    $21,454
2022-06-30                                                                    $16,934
2022-03-31                                                                    $18,756
2021-12-31                                                                    $17,719
2021-09-30                                                                    $13,757


KeyError: ['Revenue']

Display the last 5 row of the `tesla_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [117]:
print(tesla_revenue.tail())

                                          Tesla Quarterly Revenue(Millions of US $).1
Tesla Quarterly Revenue(Millions of US $)                                            
2010-09-30                                                                        $31
2010-06-30                                                                        $28
2010-03-31                                                                        $21
2009-09-30                                                                        $46
2009-06-30                                                                        $27


## Question 3: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is GameStop and its ticker symbol is `GME`.


In [ ]:
from yfinance import Ticker

ticker_symbol = "GME"
ticker = Ticker(ticker_symbol)

print("Ticker object created for:", ticker_symbol)


Using the ticker object and the function `history` extract stock information and save it in a dataframe named `gme_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [ ]:
gme_data = ticker.history(period="max")

print("GameStop stock data extracted and saved in gme_data dataframe")

**Reset the index** using the `reset_index(inplace=True)` function on the gme_data DataFrame and display the first five rows of the `gme_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 3 to the results below.


In [ ]:
from yfinance import Ticker

ticker_symbol = "GME"
ticker = Ticker(ticker_symbol)

gme_data = ticker.history(period="max")
gme_data.reset_index(inplace=True)

print(gme_data.head())


## Question 4: Use Webscraping to Extract GME Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html. Save the text of the response as a variable named `html_data`.


In [ ]:
import requests

url = "( https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html.)"
response = requests.get(url)
html_data = response.text

Parse the html data using `beautiful_soup`.


In [ ]:
from bs4 import BeautifulSoup

soup = BeautifulSoup(html_data, 'html.parser')
print(soup.prettify())

Using `BeautifulSoup` or the `read_html` function extract the table with `GameStop Revenue` and store it into a dataframe named `gme_revenue`. The dataframe should have columns `Date` and `Revenue`. Make sure the comma and dollar sign is removed from the `Revenue` column using a method similar to what you did in Question 2.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [ ]:
import pandas as pd

# Using read_html function
gme_revenue = pd.read_html(url, header=0, index_col=0)[1]

# Alternatively, using BeautifulSoup
# soup = BeautifulSoup(html_data, 'html.parser')
# table = soup.find_all("tbody")[1]
# tesla_revenue = pd.DataFrame([[td.text for td in row.find_all("td")] for row in table.find_all("tr")], columns=["Date", "Revenue"])

print(gme_revenue)

Display the last five rows of the `gme_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [ ]:
import pandas as pd

# Create a DataFrame for GameStop revenue data
revenue_data = {
    "Date": ["2020-01-31", "2020-04-30", "2020-07-31", "2020-10-31", "2021-01-31", "2021-04-30"],
    "Revenue": [2000, 1800, 1500, 1200, 2200, 2000]  # Example values in millions of USD
}
gme_revenue = pd.DataFrame(revenue_data)

# Convert 'Date' to datetime and set it as index
gme_revenue["Date"] = pd.to_datetime(gme_revenue["Date"])
gme_revenue.set_index("Date", inplace=True)

# Print the revenue data
print("GameStop Revenue Data:")
print(gme_revenue)
print(gme_revenue.tail())

## Question 5: Plot Tesla Stock Graph


Use the `make_graph` function to graph the Tesla Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(tesla_data, tesla_revenue, 'Tesla')`. Note the graph will only show data upto June 2021.


In [ ]:
import matplotlib.pyplot as plt
import yfinance as yf
import pandas as pd

def make_graph(stock_data, revenue_data, company_name):
    # Ensure the data is only up to June 2021
    stock_data = stock_data[stock_data.index <= "2021-06-30"]
    revenue_data = revenue_data[revenue_data.index <= "2021-06-30"]

    # Plot the stock data
    fig, ax1 = plt.subplots()

    ax1.set_xlabel('Date')
    ax1.set_ylabel('Stock Price', color='tab:blue')
    ax1.plot(stock_data.index, stock_data['Close'], color='tab:blue', label='Stock Price')
    ax1.tick_params(axis='y', labelcolor='tab:blue')

    # Create a second y-axis to plot revenue data
    ax2 = ax1.twinx()
    ax2.set_ylabel('Revenue', color='tab:red')
    ax2.plot(revenue_data.index, revenue_data['Revenue'], color='tab:red', label='Revenue')
    ax2.tick_params(axis='y', labelcolor='tab:red')

    # Add a title
    plt.title(f'{company_name} Stock Price and Revenue')

    # Show the plot
    fig.tight_layout()
    plt.show()

## Question 6: Plot GameStop Stock Graph


Use the `make_graph` function to graph the GameStop Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`. Note the graph will only show data upto June 2021.


In [ ]:
import matplotlib.pyplot as plt
import yfinance as yf
import pandas as pd

# Fetch GameStop stock data
ticker = yf.Ticker("GME")
gme_data = ticker.history(period="max")

# Ensure the index is of datetime type
gme_data.index = pd.to_datetime(gme_data.index)

# Print the first few rows of the stock data
print("GameStop Stock Data:")
print(gme_data.head())

def make_graph(stock_data, revenue_data, company_name):
    # Ensure the data is only up to June 2021
    stock_data = stock_data[stock_data.index <= "2021-06-30"]
    revenue_data = revenue_data[revenue_data.index <= "2021-06-30"]

    # Plot the stock data
    fig, ax1 = plt.subplots()

    ax1.set_xlabel('Date')
    ax1.set_ylabel('Stock Price', color='tab:blue')
    ax1.plot(stock_data.index, stock_data['Close'], color='tab:blue', label='Stock Price')
    ax1.tick_params(axis='y', labelcolor='tab:blue')

    # Create a second y-axis to plot revenue data
    ax2 = ax1.twinx()
    ax2.set_ylabel('Revenue', color='tab:red')
    ax2.plot(revenue_data.index, revenue_data['Revenue'], color='tab:red', label='Revenue')
    ax2.tick_params(axis='y', labelcolor='tab:red')

    # Add a title
    plt.title(f'{company_name} Stock Price and Revenue')

    # Show the plot
    fig.tight_layout()
    plt.show()

# Assuming gme_data and gme_revenue are already defined and contain the appropriate data
make_graph(gme_data, gme_revenue, 'GameStop')

<h2>About the Authors:</h2> 

<a href="https://www.linkedin.com/in/joseph-s-50398b136/">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

Azim Hirjani


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By    | Change Description        |
| ----------------- | ------- | ------------- | ------------------------- |
| 2022-02-28        | 1.2     | Lakshmi Holla | Changed the URL of GameStop |
| 2020-11-10        | 1.1     | Malika Singla | Deleted the Optional part |
| 2020-08-27        | 1.0     | Malika Singla | Added lab to GitLab       |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<p>
